systems/plants/@Manipulator/Manipulator.m

```diff
@@ -36,6 +36,73 @@
       C_times_v = C-G;
     end
     
+    function [xdot,dxdot] = Hdynamics(obj,t,x,u)
+    % Provides the DrakeSystem interface to the manipulatorDynamics.
+    %the dynamics left multiplied by H
+
+      q = x(1:obj.num_positions);
+      v = x(obj.num_positions+1:end);
+
+      if (nargout>1)
+        if ~isempty(obj.position_constraints) || ~isempty(obj.velocity_constraints)
+          % by naming this 'MATLAB:TooManyOutputs', geval will catch the
+          % error and use TaylorVarInstead
+          error('MATLAB:TooManyOutputs','User gradients for constrained dynamics not implemented yet.');
+        end
+
+        % Note: the next line assumes that user gradients are implemented.
+        % If it fails, then it will raise the same exception that I would
+        % want to raise for this method, stating that not all outputs were
+        % assigned.  (since I can't write dxdot anymore)
+        [H,C,B,dH,dC,dB] = obj.manipulatorDynamics(q,v);
+        
+        %ANDY CHANGE
+        %Hinv = inv(H);
+        %Hinv = invmsspoly(H);
+
+        if (obj.num_u>0)
+          vdot = (B*u-C);
```

Comment:
Hi Twan,

I should have documented more what that is.  Hdynamics is supposed to
return the same thing as dynamics, but with an H^-1 in front.  This was, I
thought, a useful utility function if you wanted to work in msspoly's to
get symbolic dynamics.  It is impossible to get the full dynamics, because
H cannot be inverted and kept as a polynomial expression.  The Hdynamics,
however, as I'm calling it, can still be useful for generating constraints
on dynamics.

-Andy S.

On Thu, Jul 30, 2015 at 10:55 PM, Twan Koolen <notifications@github.com>
wrote:

> In systems/plants/@Manipulator/Manipulator.m
> <https://github.com/RobotLocomotion/drake/pull/1212#discussion_r35943212>:
>
> > +          % error and use TaylorVarInstead
> > +          error('MATLAB:TooManyOutputs','User gradients for constrained dynamics not implemented yet.');
> > +        end
> > +
> > +        % Note: the next line assumes that user gradients are implemented.
> > +        % If it fails, then it will raise the same exception that I would
> > +        % want to raise for this method, stating that not all outputs were
> > +        % assigned.  (since I can't write dxdot anymore)
> > +        [H,C,B,dH,dC,dB] = obj.manipulatorDynamics(q,v);
> > +
> > +        %ANDY CHANGE
> > +        %Hinv = inv(H);
> > +        %Hinv = invmsspoly(H);
> > +
> > +        if (obj.num_u>0)
> > +          vdot = (B*u-C);
>
> The lhs would not be vdot.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/1212/files#r35943212>.
>